prow/resultstore/files.go

```diff
@@ -0,0 +1,197 @@
+	}
+
+	if o.ArtifactsDirOnly {
+		artifacts := prefix + "artifacts/"
+		match := func(name string) bool {
+			fmt.Printf("\nname: %q\n", name)
+			return name == artifacts
+		}
+		if err := c.collectDirs(ctx, prefix, match); err != nil {
+			return c.builder.Files(), err
+		}
+		return c.builder.Files(), nil
+	}
+
+	// Collect the entire artifacts/ subtree.
+	if err := c.collect(ctx, prefix+"artifacts/", ""); err != nil {
+		return c.builder.Files(), err
+	}
+	return c.builder.Files(), nil
+}
+
+func ensureTrailingSlash(p string) string {
+	if strings.HasSuffix(p, "/") {
+		return p
+	}
+	return p + "/"
+}
+
+type filesCollector struct {
+	finder  fileFinder
+	builder *filesBuilder
+}
+
+func newFilesCollector(opener fileFinder, prefix string) *filesCollector {
+	return &filesCollector{
+		finder:  opener,
+		builder: newFilesBuilder(prefix),
+	}
+}
+
+// collect collects files from storage using GCS List semantics:
+// - prefix should be a "/" terminated path.
+// - delimiter should be "/" to search a single dir
+// - delimiter should be "" to search the tree below prefix
+func (c *filesCollector) collect(ctx context.Context, prefix, delimiter string) error {
+	iter, err := c.finder.Iterator(ctx, prefix, delimiter)
+	if err != nil {
+		return err
+	}
+	for {
+		f, err := iter.Next(ctx)
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return err
+		}
+		if f.IsDir {
+			continue
+		}
+		attrs, err := c.finder.Attributes(ctx, f.Name)
+		if err != nil {
+			return err
+		}
+		c.builder.Add(f.Name, &attrs)
+	}
+	return nil
+}
+
+// collectDirs collects directories in prefix where match is true.
+func (c *filesCollector) collectDirs(ctx context.Context, prefix string, match func(string) bool) error {
+	iter, err := c.finder.Iterator(ctx, prefix, "/")
+	if err != nil {
+		return err
+	}
+	for {
+		f, err := iter.Next(ctx)
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return err
+		}
+		if !f.IsDir {
+			continue
+		}
+		if match(f.Name) {
+			c.builder.AddDir(f.Name)
+		}
+	}
+	return nil
+}
+
+type filesBuilder struct {
+	prefix string
+	trim   func(string) string
+	files  []*resultstore.File
+}
+
+func newFilesBuilder(prefix string) *filesBuilder {
+	return &filesBuilder{
+		prefix: prefix,
+		// Trims the prefix from names to produce File.Uid.
+		trim: strings.NewReplacer(prefix, "").Replace,
+	}
+}
+
+func (b *filesBuilder) Add(name string, attrs *pio.Attributes) {
+	uid := b.trim(name)
+	switch uid {
+	case "build.log":
+		// This file name is unexpected and would cause an upload
+		// exception, since ResultStore requires unique Uids.
+		return
+	case "build-log.txt":
+		// This Uid is used to populate the "Build Log" tab in the
+		// GUI. We want build-log.txt to appear there.
+		uid = "build.log"
+	}
+	b.files = append(b.files, &resultstore.File{
+		Uid:         uid,
+		Uri:         name,
+		Length:      &wrapperspb.Int64Value{Value: attrs.Size},
+		ContentType: shortContentType(attrs.ContentEncoding),
```

Comment:
>  I'm not sure what kind of behavior that would induce in ResultStore.

@ashleyrichter Do you have any insights here on the behavior of ResultStore if we omit the file length and/or the content type?

Reply:
Length is purely informational for the end user in the UI. If it's not provided, the UI will not show a length rather than try to stat all of the files referenced (as stating arbitrarily many files can be very slow).

`content_type` is used by the UI to determine how the file should be displayed, such as plain text or HTML, etc. If you don't provide the content type, the UI will try to infer it based on file extension or name. If it can't infer content type, it will default to just downloading the file rather than trying to display it in the UI (in other words, it will presume binary data for unknown files).

Reply:
Ack.

Hmm,  if we define `content_type` as plain text for everything (assume there is a bug on our end and we do this for all files, text or otherwise), the UI will attempt to read/display binary data? Seems like a possible bug and/or security issue.

In general I'm noticing a pattern of ResultStore trusting clients to always do the right thing, but this does not seem tenable in the long run. Maybe I'm missing something?

Reply:
I can point you to the internal security review with the confidential details if you reach out to me on Corp, but we treat all file data as potentially hostile and we have protections in place.

In terms of trusting the client, that's complicated and it's a problem we've spent a lot of time on. We encourage users to use a Bazel because it at least makes that problem easier to reason about, but that's not reasonable for all users. Anyways, I can talk more on that too if you reach out to me internally.

Reply:
Thanks for clarifying. I don't know what you mean by "We encourage users to use a Bazel..." (perhaps you're saying that if users all used Bazel, then at least the build artifacts could be in some way standardized, which would obviously make ResultStore's life easier).

I don't have any further questions at this time, thanks.